src/main.rs

```diff
@@ -32,128 +39,127 @@ fn main() {
 
-                            frame_capturer.run(&output_name, controller)
-                        })
-                        .unwrap_or_else(|_| panic!("Unable to start thread: {}", thread_name));
-
-                    Some(als_tx)
+                            frame_capturer.run(&output_name, controller).await;
+                        }));
+
+                        als_txs.push(als_tx);
+                    }
+                    Err(err) => {
+                        log::warn!(
+                            "Skipping '{}' as it might be disconnected: {}",
+                            output_name,
+                            err
+                        );
+                    }
                 }
-                Err(err) => {
-                    log::warn!(
-                        "Skipping '{}' as it might be disconnected: {}",
-                        output_name,
-                        err
-                    );
-
-                    None
-                }
-            }
-        })
-        .collect_vec();
-
-    std::thread::Builder::new()
-        .name("als".to_string())
-        .spawn(move || {
-            let als: Box<dyn als::Als> = match config.als {
-                config::Als::Iio { path, thresholds } => Box::new(
-                    als::iio::Als::new(&path, thresholds)
-                        .expect("Unable to initialize ALS IIO sensor"),
-                ),
-                config::Als::Time { thresholds } => Box::new(als::time::Als::new(thresholds)),
-                config::Als::Webcam { video, thresholds } => Box::new({
-                    let (webcam_tx, webcam_rx) = mpsc::channel();
-                    std::thread::Builder::new()
-                        .name("als-webcam".to_string())
-                        .spawn(move || {
-                            als::webcam::Webcam::new(webcam_tx, video).run();
-                        })
-                        .expect("Unable to start thread: als-webcam");
-                    als::webcam::Als::new(webcam_rx, thresholds)
-                }),
-                config::Als::None { .. } => Box::<als::none::Als>::default(),
-            };
-
-            als::controller::Controller::new(als, als_txs).run();
-        })
-        .expect("Unable to start thread: als");
+
+                (tasks, als_txs)
+            },
+        )
+        .await;
+
+    let als: als::Als = match config.als {
+        config::Als::Iio { path, thresholds } => als::Als::Iio(
+            als::iio::Als::new(&path, thresholds)
+                .await
+                .expect("Unable to initialize ALS IIO sensor"),
+        ),
+        config::Als::Time { thresholds } => als::Als::Time(als::time::Als::new(thresholds)),
+        config::Als::Webcam { video, thresholds } => als::Als::Webcam({
+            let (webcam_tx, webcam_rx) = channel::bounded(128);
+            tasks.push(smol::spawn({
+                // TODO: make async
+                smol::unblock(move || {
```

Comment:
We should push the return value of `unblock` to `tasks` directly and remove the outer `spawn`. I believe the types should be compatible, and it's not really doing us any favors.